### PR TITLE
dialects: Update arith, math, scf and stencil side-effect traits.

### PR DESCRIPTION
**Pull request by PapyChacal**
NB: `scf.reduce` is missing IsTerminator/HasParent traits too, but that needs updating some tests so sounded out of scope of this PR.

### CODE REVIEW

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -225,8 +229,6 @@ def __hash__(self) -> int:
 class BinaryOperationWithFastMath(Generic[_T], BinaryOperation[_T]):
     fastmath = opt_prop_def(FastMathFlagsAttr)
 
-    traits = frozenset((Pure(),))
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:228]: Why this change?

Comment (PapyChacal) [reply xdsl/dialects/arith.py:228]: Well they are not all Pure, and if we start actually using those traits, this is a timebomb!

Pure is NoMemoryEffect+AlwaysSpeculatable;
AlwaysSpeculatable implies that an operation never has undefined behaviour,
which is not the case of e.g. divisions, where dividing by 0 is UB

**Review by georgebisbas — APPROVED** [review]

File: xdsl/dialects/arith.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/arith.py:228]: Ah nice, thats very clear
